axum/src/extract/ws.rs

```diff
@@ -36,6 +36,39 @@
 //! # };
 //! ```
 //!
+//! # Passing data and/or state to an on_upgrade callback
+//!
+//! ```
+//! use std::sync::Arc;
+//!
+//! use axum::{
+//!     extract::ws::{WebSocketUpgrade, WebSocket},
+//!     response::Response,
+//!     routing::get,
+//!     Extension, Router,
+//! };
+//!
+//! struct State {
+//!     // ...
+//! }
+//!
+//! async fn handler(ws: WebSocketUpgrade, Extension(state): Extension<Arc<State>>) -> Response {
+//!     ws.on_upgrade(|socket| handle_socket(socket, state))
+//! }
+//!
+//! async fn handle_socket(socket: WebSocket, state: Arc<State>) {
+//!     // ...
+//! }
+//!
+//! let app = Router::new()
+//!     .route("/ws", get(handler))
+//!     .layer(Extension(Arc::new(State { /* ... */ })));
```

Comment:
We normally don't wrap things in `Arc` in examples to keep it simple

```suggestion
//!     .layer(Extension(State { /* ... */ }));
```